File: backend/src/Operum.API/Controllers/NotificationsController.cs
using Microsoft.AspNetCore.Mvc;
using Operum.API.Controllers.Base;
using Operum.Model.DTOs.Notifications.Requests;
using Operum.Service.Interfaces;

namespace Operum.API.Controllers
{
    [ApiController]
    [Route("api/trackers/{trackerId}/[controller]")]
    public class NotificationsController(INotificationsService notificationsService) : BaseController
    {
        [HttpGet]
        public async Task<IActionResult> GetNotifications([FromRoute] string trackerId)
        {
            return GetApiResponse(await notificationsService.GetNotifications(trackerId));
        }

        [HttpPost]
        public async Task<IActionResult> CreateNotification([FromRoute] string trackerId, [FromBody] CreateTrackerNotificationDto dto)
        {
            return GetApiResponse(await notificationsService.CreateNotification(trackerId, dto));
        }

        [HttpPut("{notificationId}")]
        public async Task<IActionResult> UpdateNotification([FromRoute] string trackerId, [FromRoute] string notificationId, [FromBody] UpdateTrackerNotificationDto dto)
        {
            return GetApiResponse(await notificationsService.UpdateNotification(trackerId, notificationId, dto));
        }

        [HttpDelete("{notificationId}")]
        public async Task<IActionResult> DeleteNotification([FromRoute] string trackerId, [FromRoute] string notificationId)
        {
            return GetApiResponse(await notificationsService.DeleteNotification(trackerId, notificationId));
        }

        [HttpPatch("{notificationId}/toggle")]
        public async Task<IActionResult> ToggleEnabled([FromRoute] string trackerId, [FromRoute] string notificationId)
        {
            return GetApiResponse(await notificationsService.ToggleEnabled(trackerId, notificationId));
        }
    }
}

File: frontend/src/features/notifications/api/notificationsController.ts
import api from "../../../shared/api/api";
import { ApiResponse } from "../../../shared/types/ApiResponse";
import { TrackerNotificationDto } from "../types/NotificationDto";
import {
    CreateTrackerNotificationDto,
    UpdateTrackerNotificationDto,
} from "../types/requests/CreateTrackerNotificationDto";

export const notificationsController = {
    getNotifications: async (
        trackerId: string
    ): Promise<ApiResponse<TrackerNotificationDto[]>> =>
        api.get(`/trackers/${trackerId}/notifications`),

    createNotification: async (
        trackerId: string,
        dto: CreateTrackerNotificationDto
    ): Promise<ApiResponse<TrackerNotificationDto>> =>
        api.post(`/trackers/${trackerId}/notifications`, dto),

    updateNotification: async (
        trackerId: string,
        notificationId: string,
        dto: UpdateTrackerNotificationDto
    ): Promise<ApiResponse<TrackerNotificationDto>> =>
        api.put(`/trackers/${trackerId}/notifications/${notificationId}`, dto),

    deleteNotification: async (
        trackerId: string,
        notificationId: string
    ): Promise<ApiResponse> =>
        api.delete(`/trackers/${trackerId}/notifications/${notificationId}`),

    toggleEnabled: async (
        trackerId: string,
        notificationId: string
    ): Promise<ApiResponse<TrackerNotificationDto>> =>
        api.patch(`/trackers/${trackerId}/notifications/${notificationId}/toggle`, {}),
};
